cfme/utils/appliance/implementations/ui.py

```diff
@@ -174,6 +203,13 @@ def before_keyboard_input(self, element, keyboard_input):
         sleep(0.3)
         self.make_document_focused()
 
+    def before_click(self, element):
+        # this is necessary in order to handle unexpected alerts like "Abandon Changes"
+        self.browser.page_dirty = self.page_has_changes
+
+    def after_click(self, element):
+        self.browser.page_dirty = None
```

Comment:
Why set to `None` instead of call self.page_has_changes again? Its possible the click only changed part of a page.

Reply:
I would imagine to not have to call it again - but it's a good point

Reply:
it is necessary to check page for changes only before doing something now. Because we need to close "Abandon changes" alert. 

There is a reason to set page_dirty to None or False in after_click. If it was left True, all next calls of ensure_page_safe would check for alert presence. whereas alert methods are enormously slow in selenium. 
btw, I can set it to None/False immediately after closing alert if you wish.
btw, I can also update page_dirty after any manipulation in selenium but it will slow down our tests.

Reply:
Please add some inline comments concerning what you've found @izapolsk, thanks!